aten/src/ATen/core/ivalue.h

```diff
@@ -458,13 +457,21 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
     if (completed()) {
       return;
     }
-    c10::global_work_queue.workOnTasksUntilCompleted(intrusive_from_this());
+    c10::global_work_queue.waitWorkComplete(intrusive_from_this());
     AT_ASSERT(completed());
   }
 
   void markCompleted(IValue value) {
-    value_ = std::move(value);
-    completed_ = true;
+    {
+      // This is not to protect completed_ but to create a barrier
+      // from possible addCallback() calls
+      std::unique_lock<std::mutex> lock(mutex_);
+      AT_ASSERT(!completed());
+      completed_ = true;
+      value_ = std::move(value);
+    }
+
+    // no need to protect callbacks anymore
```

Comment:
This comment doesn't explain why

Reply:
Shouldn't we clear the callbacks after they are run?

Reply:
This is what `callbacks.clear();` does?